halo2/src/recursion/endoscale/primitive.rs

```diff
@@ -0,0 +1,161 @@
+//! Primitives used in endoscaling.
+
+use group::{Curve, Group};
+use pasta_curves::arithmetic::{CurveAffine, FieldExt};
+
+use subtle::CtOption;
+
+/// Maps a pair of bits to a multiple of a scalar using endoscaling.
+pub(crate) fn endoscale_pair_scalar<F: FieldExt>(bits: [bool; 2]) -> F {
+    // [2 * bits.0 - 1]
+    let mut scalar = F::from(bits[0]).double() - F::one();
+
+    if bits[1] {
+        scalar *= F::ZETA;
+    }
+
+    scalar
+}
+
+/// Maps a K-bit bitstring to a scalar.
+///
+/// This corresponds to Algorithm 1 from [BGH2019], where `F` corresponds to $F_q$, the
+/// scalar field of $P$. Where Algorithm 1 computes $Acc = [scalar] P$, this function
+/// computes `scalar`.
+///
+/// [BGH2019]: https://eprint.iacr.org/2019/1021.pdf
+#[allow(dead_code)]
+pub(crate) fn endoscale_scalar<F: FieldExt>(acc: Option<F>, bits: &[bool]) -> F {
```

Comment:
This API takes an optional `acc`, which Algorithm 2 does not; this feels like the function is only partially implementing Algorithm 2 in that case. Document what is going on here when `acc.is_some()`.

EDIT: @therealyingtong explained that `acc` is used specifically for how the efficient chip implements this using a lookup table. That is implementation-specific, but this crate function doesn't live under `endoscale::chip` which is what was confusing me. Rename this function to clarify its purpose, and then have a separate `endoscale_scalar` function that is documented like this currently is (just Alg 2), and does not take `acc` as an argument.